PWX-27654: CSR support

zoxpx opened the pull request:
Signed-off-by: Zoran Rajic <zrajic@purestorage.com>

<!--
  Make sure to have done the following:
  [] Signed off your work as per the DCO.
  [] Add unit-tests
-->

**What this PR does / why we need it**:
This change adds support for Kubernetes CSR (Certificate Signing Requests).

Additionally, adding the `UpdateEvent()` to allow updating pre-existing K8s Events

**Which issue(s) this PR fixes** (optional)
Closes # PWX-27654

**Special notes for your reviewer**:
Required for "auto-SSL" setup of Portworx in K8s environments.

piyush-nimbalkar commented on k8s/core/core.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/portworx/sched-ops/k8s/common"
 	"github.com/portworx/sched-ops/task"
 	"github.com/sirupsen/logrus"
+	certv1 "k8s.io/api/certificates/v1"
```
We have been creating separate clients under k8s folder for different API groups. Core, rbac, policy, etc.
Let's do the same for certificate APIs for consistency?

zoxpx replied:
> We have been creating separate clients under k8s folder for different API groups. Core, rbac, policy, etc. Let's do the same for certificate APIs for consistency?

@piyush-nimbalkar -- not sure what you are asking me to do here.  Can you give me an example?

Note that I had do add small changes into `core.go`, to include the CSR interface into the main sched-ops client, and also to handle the CSR watches.  The implementation went into a separate `csr.go` file.

piyush-nimbalkar replied:
I meant move it to a separate package just like core and not just to a separate file - `k8s/certificates/csr.go`
We did that for all the different k8s apis.

zoxpx replied:
> We did that for all the different k8s apis

I checked the [core.handleWatch()](https://github.com/portworx/sched-ops/blob/master/k8s/core/core.go#L225) -- and `WatchNode()`, `WatchConfigMap()`, `WatchPods()` and `WatchSecret()` are all implemented in the `core` package.
* since `Certificates` is IMO a concept very close to `Secrets` -- I followed how/where the `Secrets` API was implemented, and thus added it into the `core` package
* on the other hand, k8s-subdirs seem to be product-related  (`k8s/{stork,talisman,openshift,prometheus,autopilot}`...)

Is this going to be a problem?

piyush-nimbalkar replied:
> k8s-subdirs seem to be product-related

That's not true. We have packages like rbac, apps, storage, batch, policy, dynamic, etc. All of them are separated by the Kubernetes API. Just because the concept is close to Secrets does not mean we should make an exception.

Of course, it is not going to a problem because this won't vendor new outside packages (one of the reasons why we separated these packages). It is just about consistency.

I have approved the PR already, so will leave it up to you. Just wanted to point out the inconsistency.

zoxpx commented:
**UPDATE**: Sorry, didn't notice this before -- the code-comments were a bit off-sync.  Added follow-up [7e490ee](https://github.com/portworx/sched-ops/pull/290/commits/7e490eee5ab5f58044c072a446ef9f87706511e5) to fix.

zoxpx commented:
> **UPDATE**: Sorry, didn't notice this before -- the code-comments were a bit off-sync. Added follow-up [7e490ee](https://github.com/portworx/sched-ops/pull/290/commits/7e490eee5ab5f58044c072a446ef9f87706511e5) to fix.

Sorry, I forgot to add DSO to the commit, which broke Travis validation.   Fixed GIT comment + re-pushed the commit

zoxpx commented:
Ok.. I have to make progress and start integrating this PR.   There's plenty more work that needs to follow up using this code (i.e. PX-Operator integration), so there will be opportunity to reorganize and optimize (also "battle-test") this code.

Thanks for the review Piyush
